### PR TITLE
VC look up preformance optimization

### PR DESCRIPTION
**Pull request by ztiffany**
Improve VC lookup performance. 

@hppritcha @sungeunchoi @e-harvey 

Updated vc_lookup criterion test results without performance changes:

tiger$ cat test.out.novcperf
type: MAP   npeers:       2 naddrs:       2 ns: 118.400000
type: MAP   npeers:      32 naddrs:       2 ns: 121.600000
type: MAP   npeers:      32 naddrs:      32 ns: 107.400000
type: MAP   npeers:     512 naddrs:       2 ns: 116.600000
type: MAP   npeers:     512 naddrs:      32 ns: 138.700000
type: MAP   npeers:     512 naddrs:     512 ns: 158.100000
type: MAP   npeers:    8192 naddrs:       2 ns: 108.300000
type: MAP   npeers:    8192 naddrs:      32 ns: 127.300000
type: MAP   npeers:    8192 naddrs:     512 ns: 215.400000
type: MAP   npeers:    8192 naddrs:    8192 ns: 194.500000
type: MAP   npeers:  131072 naddrs:       2 ns: 106.800000
type: MAP   npeers:  131072 naddrs:      32 ns: 221.900000
type: MAP   npeers:  131072 naddrs:     512 ns: 416.400000
type: MAP   npeers:  131072 naddrs:    8192 ns: 543.200000
type: MAP   npeers:  131072 naddrs:  131072 ns: 255.500000
type: TABLE npeers:       2 naddrs:       2 ns: 20.100000
type: TABLE npeers:      32 naddrs:       2 ns: 21.900000
type: TABLE npeers:      32 naddrs:      32 ns: 21.300000
type: TABLE npeers:     512 naddrs:       2 ns: 19.500000
type: TABLE npeers:     512 naddrs:      32 ns: 19.600000
type: TABLE npeers:     512 naddrs:     512 ns: 19.600000
type: TABLE npeers:    8192 naddrs:       2 ns: 19.400000
type: TABLE npeers:    8192 naddrs:      32 ns: 20.200000
type: TABLE npeers:    8192 naddrs:     512 ns: 20.000000
type: TABLE npeers:    8192 naddrs:    8192 ns: 20.200000
type: TABLE npeers:  131072 naddrs:       2 ns: 44.600000
type: TABLE npeers:  131072 naddrs:      32 ns: 104.300000
type: TABLE npeers:  131072 naddrs:     512 ns: 262.400000
type: TABLE npeers:  131072 naddrs:    8192 ns: 292.700000
type: TABLE npeers:  131072 naddrs:  131072 ns: 20.500000

With performance changes:

tiger$ cat test.out.vcperf
type: MAP   npeers:       2 naddrs:       2 ns: 47.800000
type: MAP   npeers:      32 naddrs:       2 ns: 41.100000
type: MAP   npeers:      32 naddrs:      32 ns: 42.100000
type: MAP   npeers:     512 naddrs:       2 ns: 37.800000
type: MAP   npeers:     512 naddrs:      32 ns: 46.000000
type: MAP   npeers:     512 naddrs:     512 ns: 69.100000
type: MAP   npeers:    8192 naddrs:       2 ns: 34.100000
type: MAP   npeers:    8192 naddrs:      32 ns: 44.300000
type: MAP   npeers:    8192 naddrs:     512 ns: 77.700000
type: MAP   npeers:    8192 naddrs:    8192 ns: 84.200000
type: MAP   npeers:  131072 naddrs:       2 ns: 36.400000
type: MAP   npeers:  131072 naddrs:      32 ns: 66.200000
type: MAP   npeers:  131072 naddrs:     512 ns: 175.500000
type: MAP   npeers:  131072 naddrs:    8192 ns: 236.700000
type: MAP   npeers:  131072 naddrs:  131072 ns: 103.600000
type: TABLE npeers:       2 naddrs:       2 ns: 25.800000
type: TABLE npeers:      32 naddrs:       2 ns: 26.900000
type: TABLE npeers:      32 naddrs:      32 ns: 26.900000
type: TABLE npeers:     512 naddrs:       2 ns: 25.000000
type: TABLE npeers:     512 naddrs:      32 ns: 25.000000
type: TABLE npeers:     512 naddrs:     512 ns: 25.300000
type: TABLE npeers:    8192 naddrs:       2 ns: 25.100000
type: TABLE npeers:    8192 naddrs:      32 ns: 27.200000
type: TABLE npeers:    8192 naddrs:     512 ns: 25.700000
type: TABLE npeers:    8192 naddrs:    8192 ns: 27.200000
type: TABLE npeers:  131072 naddrs:       2 ns: 25.000000
type: TABLE npeers:  131072 naddrs:      32 ns: 30.500000
type: TABLE npeers:  131072 naddrs:     512 ns: 38.300000
type: TABLE npeers:  131072 naddrs:    8192 ns: 45.600000
type: TABLE npeers:  131072 naddrs:  131072 ns: 26.200000
[====] Synthesis: Tested: 1 | Passing: 1 | Failing: 0 | Crashing: 0

### CODE REVIEW

**Comment by ztiffany** [comment]
I have run the following test:
1. Criterion tests
2. cray_runall.sh suite
3. osu_latency+ompi over GNI

Let me know if there are more tests I should do.  I'm a little unsure if I've got all connection setup paths covered.

File: prov/gni/src/gnix_vc.c [hunk 1]
```diff
@@ -86,6 +86,94 @@ static inline void __gnix_vc_set_ht_key(void *gnix_addr,
 	*key = *((gnix_ht_key_t *)gnix_addr);
 }
 
+struct gnix_vc *_gnix_ep_vc_lookup(struct gnix_fid_ep *ep, uint64_t key)
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_vc.c]: This should be static.

File: prov/gni/src/gnix_vc.c [hunk 2]
```diff
@@ -86,6 +86,94 @@ static inline void __gnix_vc_set_ht_key(void *gnix_addr,
...
+			vc = NULL;
+		}
+	} else {
+		vc = (struct gnix_vc *)_gnix_ht_lookup(ep->vc_ht, key);
+	}
+
+	return vc;
+}
+
+int _gnix_ep_vc_store(struct gnix_fid_ep *ep, struct gnix_vc *vc, uint64_t key)
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_vc.c]: This should be static.

File: prov/gni/src/gnix_vc.c [hunk 3]
```diff
@@ -86,6 +86,94 @@ static inline void __gnix_vc_set_ht_key(void *gnix_addr,
...
+					 (void *)&av_entry->gnix_addr);
+	if (entry) {
+		/* Found a matching, unmapped VC.  Map dest_addr to the VC in
+		 * the EP's VC look up table. */
+		vc = dlist_entry(entry, struct gnix_vc, list);
+		GNIX_INFO(FI_LOG_EP_CTRL,
+			  "Found unmapped VC: %p gnix_addr: 0x%lx fi_addr: 0x%lx\n",
+			  vc, vc->peer_addr, vc->peer_fi_addr);
+
+		ret = _gnix_ep_vc_store(ep, vc, dest_addr);
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_vc.c:163]: Under what conditions would this operation fail?

**Comment by sungeunchoi** [comment]
You could also run the OSU suite as we do with our jenkins testing with https://github.com/ofi-cray/fab-utils/blob/master/scripts/benchmarks/run_osu.

File: prov/gni/src/gnix_ep.c [hunk 4]
```diff
@@ -272,77 +273,123 @@ static inline int __gnix_ep_init_vc(struct gnix_fid_ep *ep_priv)
...
 		if(ep_priv->vc_table == NULL)
-			goto err;
+			return -FI_ENOMEM;
 
-		gnix_vec_attr.vec_initial_size = ep_priv->domain->params.ct_init_size;
-		gnix_vec_attr.vec_maximum_size = ep_priv->domain->params.ct_max_size;
+		gnix_vec_attr.vec_initial_size =
+				ep_priv->domain->params.ct_init_size;
+		/* TODO: ep_priv->domain->params.ct_max_size; */
+		gnix_vec_attr.vec_maximum_size = 1024*1024;
```
Comment (e-harvey) [inline prov/gni/src/gnix_ep.c:285]: Should the vec_maximum_size be set to ep_priv->domain->params.ct_max_size here?

**Comment by jswaro** [comment]
How consistent are the results? The change is really nice. Is there much variance in the results?

File: prov/gni/src/gnix_vc.c [hunk 5]
```diff
@@ -86,6 +86,94 @@ static inline void __gnix_vc_set_ht_key(void *gnix_addr,
...
+}
+
+int _gnix_ep_vc_store(struct gnix_fid_ep *ep, struct gnix_vc *vc, uint64_t key)
+{
+	int ret;
+
+	assert(ep->av);
+
+	if (ep->av->type == FI_AV_TABLE) {
+		ret = _gnix_vec_insert_at(ep->vc_table, (void *)vc, key);
```
Comment (e-harvey) [inline prov/gni/src/gnix_vc.c:116]: This needs vc_modes |= GNIX_VC_MODE_IN_TABLE

File: prov/gni/src/gnix_vc.c [hunk 6]
```diff
@@ -86,6 +86,94 @@ static inline void __gnix_vc_set_ht_key(void *gnix_addr,
...
+	} else {
+		ret = _gnix_ht_insert(ep->vc_ht, key, vc);
```
Comment (e-harvey) [inline prov/gni/src/gnix_vc.c:118]: This needs vc_modes |= GNIX_VC_MODE_IN_HT

File: prov/gni/src/gnix_vc.c [hunk 7]
```diff
@@ -96,143 +184,88 @@ static inline void __gnix_vc_set_ht_key(void *gnix_addr,
...
 			goto err_w_lock;
 		}
-	} else  {
 
-		/*
-		 * check here if the vc should be in the table accelerated
-		 * lookup but hasn't yet been inserted
-		 */
+		vc = vc_tmp;
+		vc->modes |= GNIX_VC_MODE_IN_HT;
```
Comment (e-harvey) [inline prov/gni/src/gnix_vc.c]: This should be removed.

**Comment by sungeunchoi** [comment]
Are these VC_MODE bits even needed anymore?  The vector one is not used, and the hashtable one is only used for sanity checking (which you might be able to in a different way).

File: prov/gni/src/gnix_vc.c (hunk 3 above)
Comment (ztiffany) [inline prov/gni/src/gnix_vc.c:163]: I had seen this fail when using vector storage and hitting the upper limit.  The vector enforces a hard limit of entries.  The hashtable doesn't have a hard limit.  We might need to remove the vector size limit.  I added an arbitrarily large limit for now.  This shouldn't fail under normal conditions.

**Comment by ztiffany** [comment]
@jswaro results are pretty consistent.

@sungeunchoi @e-harvey I can make the requested changes to the flags.  I don't think they ever served a function, though.  I assume they are for debugging?

File: prov/gni/src/gnix_vc.c [hunk 8]
```diff
@@ -96,143 +184,88 @@ static inline void __gnix_vc_set_ht_key(void *gnix_addr,
...
+	COND_RELEASE(ep->requires_lock, &ep->vc_lock);
+
 	*vc_ptr = vc;
 	return ret;
 
 err_w_lock:
-	fastlock_release(&ep->vc_lock);
-err:
+	COND_RELEASE(ep->requires_lock, &ep->vc_lock);
 	if (vc != NULL)
```
Comment (e-harvey) [inline prov/gni/src/gnix_vc.c:267]: Do we need to check to see if this vc has been stored in the connection table and move it back to the unmapped vc cache?  Maybe just move the vc from the vector/hashtable into the unmapped vc cache on line 258, if we fail to connect?

File: prov/gni/src/gnix_vc.c [hunk 9]
```diff
@@ -921,20 +944,32 @@ static int __gnix_vc_hndl_conn_req(struct gnix_cm_nic *cm_nic,
...
-				vc = vc_try;
-				vc->modes |= GNIX_VC_MODE_IN_HT;
+			if (src_mapped) {
+				/* We have an AV which maps the incoming
+				 * address.  Store the new VC in our VC lookup
+				 * table. */
+				ret = _gnix_ep_vc_store(ep, vc_try, fi_addr);
+				if (likely(ret == FI_SUCCESS)) {
+					vc = vc_try;
+					vc->modes |= GNIX_VC_MODE_IN_HT;
```
Comment (e-harvey) [inline prov/gni/src/gnix_vc.c]: This needs to be removed.

File: prov/gni/src/gnix_vc.c (hunk 8 above)
Comment (ztiffany) [inline prov/gni/src/gnix_vc.c:267]: I don't think that's necessary right now.  If we find an unmapped VC we only unlock and return (skipping the whole 'if (vc == NULL) block).  There's no other way to fail.

**Comment by hppritcha** [comment]
I'm okay with this once feedback has been responded to.

**Pull request merged**